service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -482,4 +482,58 @@ public void registerLivenessChecks(final HealthCheckHandler handler) {
             status.complete(Status.OK());
         });
     }
+    private Future<CredentialsObject> getCredentialsForDevice(final String tenantId, final String type, final String authId) {
+
+        Future<CredentialsObject> result = Future.future();
+        getCredentialsClient(tenantId).compose(client -> {
+            Future<CredentialsResult<CredentialsObject>> credResultFuture = Future.future();
+            client.get(type, authId, credResultFuture.completer());
+            return credResultFuture;
+        }).compose(credResult -> {
+            if (credResult.getStatus() == HttpURLConnection.HTTP_OK) {
+                CredentialsObject payload = credResult.getPayload();
+                result.complete(payload);
+            } else if (credResult.getStatus() == HttpURLConnection.HTTP_NOT_FOUND) {
+                result.fail(String.format("cannot retrieve credentials (not found for type <%s>, authId <%s>)",type,authId));
+            } else {
+                result.fail("cannot retrieve credentials");
+            }
+        }, result);
+
+        return result;
+    }
+
+    /**
+     * Validates an authentication object against credentials secrets available by the get operation of the
+     *  <a href="https://www.eclipse.org/hono/api/Credentials-API/">Credentials API</a>.
+     * <p>The credentials are first retrieved from the credentials service, and then all matching validators are
+     * invoked until one is successful or all failed. The authentication object is validated iff at least
+     * one validator was successful.
+     *
+     * @param tenantId The tenantId to which the device belongs.
+     * @param type The type of credentials that are to be used for validation.
+     * @param authId The authId of the credentials that are to be used for validation.
+     * @param authenticationObject The authentication object to be validated, e.g. a password, a preshared-key, etc.
+
+     * @return Future The future object carrying the payload of the credentials get operation, if successful.
+     */
+    protected Future<String> validateCredentialsForDevice(final String tenantId, final String type, final String authId,
+                                                              final Object authenticationObject) {
+        return getCredentialsForDevice(tenantId, type, authId).compose(payload -> {
+            Future<String> resultDeviceId = Future.future();
+            CredentialsSecretsValidator validator = CredentialsUtils.findAppropriateValidators(type);
+            try {
+                if (validator != null && validator.validate(payload, authenticationObject)) {
+                    resultDeviceId.complete(payload.getDeviceId());
+                } else {
+                    resultDeviceId.fail("credentials invalid - not validated");
+                }
+            }
+            catch (IllegalArgumentException e) {
```

Comment:
I commented it to keep the exception, but I think both ways are fine for me.
Important is for me the tracing of wrong credentials, so I added a Logger to the small helper class (which was previously carried inside the exception).
Then the return of null works.